Type annotation requires UDFs to return a type instead of an array

**Describe the bug**
Here:

https://github.com/apache/datafusion-python/blob/e0284c6e788b6fc893495ed929b9badef1cf925c/python/datafusion/user_defined.py#L36

we bind `_R` to be a subtype of `pyarrow.DataType`

And in these two places we require that functions return a value of type `_R` (ie. they should return a pyarrow datatype):

https://github.com/apache/datafusion-python/blob/e0284c6e788b6fc893495ed929b9badef1cf925c/python/datafusion/user_defined.py#L123-L130

https://github.com/apache/datafusion-python/blob/e0284c6e788b6fc893495ed929b9badef1cf925c/python/datafusion/user_defined.py#L166-L174

which seems incorrect, as functions should return *arrays* that contain values of that type

**To Reproduce**

```
datafusion-python ((53.0.0)) $ mypy examples/python-udf.py
examples/python-udf.py:27: error: Value of type variable "_R" of function cannot be "Array[Any]"  [type-var]
Found 1 error in 1 file (checked 1 source file)
```

**Expected behavior**
mypy should pass

**Additional context**

--- a/python/datafusion/user_defined.py
+++ b/python/datafusion/user_defined.py
@@ -33,7 +33,7 @@ from datafusion.expr import Expr
 if TYPE_CHECKING:
     from _typeshed import CapsuleType as _PyCapsule
 
-    _R = TypeVar("_R", bound=pa.DataType)
+    _R = TypeVar("_R", bound=pa.Array)
     from collections.abc import Callable, Sequence
 
 
@@ -125,7 +125,7 @@ class ScalarUDF:
         name: str,
         func: Callable[..., _R],
         input_fields: list[pa.Field],
-        return_field: _R,
+        return_field: pa.Field,
         volatility: Volatility | str,
     ) -> None:
         """Instantiate a scalar user-defined function (UDF).
@@ -264,7 +264,7 @@ class ScalarUDF:
 
         def _decorator(
             input_fields: Sequence[pa.DataType | pa.Field] | pa.DataType | pa.Field,
-            return_field: _R,
+            return_field: pa.DataType | pa.Field,
             volatility: Volatility | str,
             name: str | None = None,
         ) -> Callable:

